File: src.swift/ImportChrome/Sources/ImportChrome/ImportChrome.swift
// Copyright (C) 2026 longtimeno-c
// SPDX-License-Identifier: GPL-3.0-or-later
// Import.bundle's principal class (plan/18 "Mac chrome"): the host
// (src/shell/addons_mac.mm) instantiates it with NSBundle and talks to it by
// message send. It owns the one Import window and every running job, so
// closing the window keeps an import running with a line in the command bar.
import AppKit
import CImportApi
import SwiftUI

@objc(MVImportChrome)
public final class MVImportChrome: NSObject {
  private var table: ImportTable?
  private var host: NSObject?
  private var model: ImportModel?
  private var window: NSWindow?
  private var timer: Timer?
  private var jobs: [UInt64: String] = [:]

  @objc public override init() { super.init() }

  @objc(attachWithTable:host:)
  public func attach(table: NSValue, host: NSObject) {
    guard let raw = table.pointerValue else { return }
    self.table = ImportTable(raw.assumingMemoryBound(to: mv_import_api.self))
    self.host = host
  }

  @MainActor private func ensureModel() -> ImportModel? {
    if let model { return model }
    guard let table else { return nil }
    let m = ImportModel(table: table)
    m.chrome = self
    model = m
    return m
  }

  @objc(openWithSource:marks:)
  public func open(source: String, marks: [String]) {
    MainActor.assumeIsolated {
      guard let model = ensureModel() else { return }
      model.marks = marks
      if window == nil {
        let w = NSWindow(contentRect: NSRect(x: 0, y: 0, width: 1280, height: 760),
                         styleMask: [.titled, .closable, .resizable, .miniaturizable],
                         backing: .buffered, defer: false)
        w.title = "Import"
        w.isReleasedWhenClosed = false
        w.contentView = NSHostingView(rootView: ImportView(model: model))
        w.center()
        window = w
      }
      window?.makeKeyAndOrderFront(nil)
      model.refreshSources(select: source.isEmpty ? nil : source)
      model.checkUnfinished()
    }
  }

  @objc(importNow:)
  public func importNow(_ pathsJSON: String) {
    guard let table else { return }
    var id: UInt64 = 0
    if table.api.pointee.import_now!(table.ctx, pathsJSON, &id) == MV_OK { track(id, label: "marked files") }
  }

  @objc(deliverEvent:status:identifier:payload:)
  public func deliver(event: UInt32, status: UInt32, identifier: UInt64, payload: Int64) {
    MainActor.assumeIsolated {
      switch event {
      case MV_ADDON_EVENT_VOLUME_ARRIVED.rawValue:
        if payload == Int64(MV_IMPORT_ARRIVAL_AUTO.rawValue) {
          track(identifier, label: "auto-import")
        } else if payload == Int64(MV_IMPORT_ARRIVAL_OPEN.rawValue), let table {
          let root = table.path { table.api.pointee.arrival_root!(table.ctx, identifier, $0, $1) } ?? ""
          if !root.isEmpty { open(source: root, marks: model?.marks ?? []) }
        }
        model?.refreshSources(select: nil)
      case MV_ADDON_EVENT_VOLUME_REMOVED.rawValue:
        model?.refreshSources(select: nil)
      case MV_ADDON_EVENT_SCAN_DONE.rawValue:
        if identifier == 0 { model?.refreshSources(select: nil) } else { model?.scanDone(identifier, status) }
      case MV_ADDON_EVENT_PLAN_READY.rawValue:
        model?.planReady(identifier, status)
      case MV_ADDON_EVENT_JOB_PROGRESS.rawValue:
        tick()
      case MV_ADDON_EVENT_JOB_DONE.rawValue, MV_ADDON_EVENT_VERIFY_DONE.rawValue:
        tick()
        finished(identifier, state: UInt32(truncatingIfNeeded: payload))
      default:
        break
      }
    }
  }

  @objc public func shutdown() {
    MainActor.assumeIsolated {
      timer?.invalidate()
      timer = nil
      window?.close()
      window = nil
      model = nil
      setStatus(nil)
      table = nil
    }
  }

  // MARK: host services (selectors on the host object, addons_mac.mm)

  func openInViewer(_ path: String) {
    host?.perform(NSSelectorFromString("openInViewer:"), with: path as NSString)
  }

  private func setStatus(_ text: String?) {
    host?.perform(NSSelectorFromString("setStatus:"), with: (text ?? "") as NSString)
  }

  func track(_ job: UInt64, label: String) {
    jobs[job] = label
    if timer == nil {
      timer = Timer.scheduledTimer(withTimeInterval: 0.25, repeats: true) { [weak self] _ in
        MainActor.assumeIsolated { self?.tick() }
      }
    }
    tick()
  }

  private func tick() {
    guard let table else { return }
    var line: String?
    for job in jobs.keys {
      var p = mv_import_progress()
      guard table.api.pointee.progress!(table.ctx, job, &p) == MV_OK else { jobs[job] = nil; continue }
      MainActor.assumeIsolated { model?.progressed(job, p) }
      if p.state == MV_IMPORT_JOB_RUNNING.rawValue || p.state == MV_IMPORT_JOB_PAUSED.rawValue ||
          p.state == MV_IMPORT_JOB_QUEUED.rawValue {
        let pct = p.bytes_total == 0 ? 0 : Int(100 * Double(p.bytes_read) / Double(p.bytes_total))
        line = (p.state == MV_IMPORT_JOB_PAUSED.rawValue ? "Import paused · " : "Importing · ") + "\(pct)%"
      }
    }
    setStatus(line)
    if jobs.isEmpty { timer?.invalidate(); timer = nil }
  }

  private func finished(_ job: UInt64, state: UInt32) {
    MainActor.assumeIsolated { model?.finished(job) }
    guard jobs.removeValue(forKey: job) != nil else { return }
    let text: String
    switch state {
    case MV_IMPORT_JOB_DONE.rawValue: text = "Finished. Every copy verified."
    case MV_IMPORT_JOB_FAILED.rawValue: text = "Finished with files that could not be copied."
    case MV_IMPORT_JOB_INTERRUPTED.rawValue: text = "Interrupted. Reconnect and resume in Import."
    default: text = "Cancelled. Nothing partial was left behind."
    }
    host?.perform(NSSelectorFromString("notifyTitle:body:"), with: "Import" as NSString, with: text as NSString)
    if jobs.isEmpty { setStatus(nil) }
  }
}

File: src.managed/MediaViewer.Import.Chrome/ImportChrome.cs
// Copyright (C) 2026 longtimeno-c
// SPDX-License-Identifier: GPL-3.0-or-later
using System.Text.Json;
using MediaViewer.Interop;
using Microsoft.UI.Dispatching;

namespace MediaViewer.Import.Chrome;

/// <summary>
/// The add-on chrome's entry point (plan/18 "Windows chrome"). Owns the one
/// Import window and every running job's progress, so closing the window
/// keeps an import running with a small indicator in the main command bar,
/// and a notification when it finishes.
/// </summary>
public sealed class ImportChrome : IAddonChrome
{
    private IAddonHost? _host;
    private ImportApi? _api;
    private ImportWindow? _window;
    private DispatcherQueueTimer? _timer;
    private readonly HashSet<ulong> _jobs = new();
    private readonly Dictionary<ulong, string> _jobLabels = new();

    internal ImportApi Api => _api ?? throw new InvalidOperationException("not attached");
    internal IAddonHost Host => _host ?? throw new InvalidOperationException("not attached");

    public void Attach(IAddonHost host, IntPtr interfaceTable)
    {
        _host = host;
        _api = new ImportApi(interfaceTable);
        _timer = DispatcherQueue.GetForCurrentThread()?.CreateTimer();
        if (_timer is not null)
        {
            _timer.Interval = TimeSpan.FromMilliseconds(250);
            _timer.Tick += (_, _) => Tick();
        }
    }

    public void Open(string? sourceRoot)
    {
        if (_api is null || _host is null) return;
        if (_window is null)
        {
            _window = new ImportWindow(this);
            _window.Closed += (_, _) => _window = null;
        }
        _window.Show(sourceRoot, _host.MarkedPaths());
    }

    public void ImportNow(string pathsJson)
    {
        if (_api is null) return;
        try
        {
            ulong job = _api.ImportNow(pathsJson);
            Track(job, "marked files");
        }
        catch (MediaViewerException ex)
        {
            _host?.SetStatus("Import could not start: " + ex.Status);
        }
    }

    internal void Track(ulong job, string label)
    {
        _jobs.Add(job);
        _jobLabels[job] = label;
        _timer?.Start();
        Tick();
    }

    public void OnEvent(AddonCompletion e)
    {
        switch (e.Event)
        {
            case MvAddonEvent.VolumeArrived:
                if ((MvImportArrival)e.Payload == MvImportArrival.Auto)
                {
                    Track(e.Id, "auto-import");
                }
                else if ((MvImportArrival)e.Payload == MvImportArrival.Open)
                {
                    string root = Try(() => Api.ArrivalRoot(e.Id)) ?? "";
                    if (root.Length > 0) Open(root);
                }
                _window?.OnSourcesChanged();
                break;
            case MvAddonEvent.VolumeRemoved:
            case MvAddonEvent.ScanDone when e.Id == 0:
                _window?.OnSourcesChanged();
                break;
            case MvAddonEvent.ScanDone:
                _window?.OnScanDone(e.Id, e.Status);
                break;
            case MvAddonEvent.PlanReady:
                _window?.OnPlanReady(e.Id, e.Status);
                break;
            case MvAddonEvent.JobProgress:
                Tick();
                break;
            case MvAddonEvent.JobDone:
            case MvAddonEvent.VerifyDone:
                Tick();
                Finished(e.Id, (MvImportJobState)(uint)e.Payload);
                break;
        }
    }

    private void Finished(ulong job, MvImportJobState state)
    {
        bool mine = _jobs.Remove(job);
        _window?.OnJobDone(job, state);
        if (!mine || _host is null) return;
        string label = _jobLabels.GetValueOrDefault(job, "Import");
        string text = state switch
        {
            MvImportJobState.Done => "Finished. Every copy verified.",
            MvImportJobState.Failed => "Finished with files that could not be copied. Open Import to retry them.",
            MvImportJobState.Interrupted => "Interrupted. Reconnect the card or drive and resume in Import.",
            _ => "Cancelled. Nothing partial was left behind.",
        };
        try
        {
            using JsonDocument s = JsonDocument.Parse(Api.SummaryJson(job));
            if (s.RootElement.TryGetProperty("copied", out JsonElement copied) &&
                copied.TryGetProperty("files", out JsonElement files))
            {
                text = $"{files.GetInt64()} files copied. " + text;
            }
        }
        catch (Exception ex) when (ex is MediaViewerException or JsonException) { }
        _host.Notify("Import: " + label, text);
        if (_jobs.Count == 0)
        {
            _timer?.Stop();
            _host.SetStatus(null);
        }
    }

    private void Tick()
    {
        if (_api is null || _host is null) return;
        string? line = null;
        foreach (ulong job in _jobs.ToArray())
        {
            MvImportProgress p;
            try { p = _api.Progress(job); }
            catch (MediaViewerException) { _jobs.Remove(job); continue; }
            _window?.OnProgress(job, p);
            if (p.State is MvImportJobState.Running or MvImportJobState.Paused or MvImportJobState.Queued)
            {
                double pct = p.BytesTotal == 0 ? 0 : 100.0 * p.BytesRead / p.BytesTotal;
                line = p.State == MvImportJobState.Paused
                    ? $"Import paused · {pct:0}%"
                    : $"Importing · {pct:0}%" + (p.EtaSeconds >= 0 ? $" · {Format.Eta(p.EtaSeconds)}" : "");
            }
        }
        _host.SetStatus(line);
        if (_jobs.Count == 0) _timer?.Stop();
    }

    public void Shutdown()
    {
        _timer?.Stop();
        _window?.Close();
        _window = null;
        _api = null;
        _host?.SetStatus(null);
        _host = null;
    }

    internal static T? Try<T>(Func<T> f) where T : class
    {
        try { return f(); }
        catch (MediaViewerException) { return null; }
    }
}

internal static class Format
{
    public static string Bytes(long b)
    {
        string[] units = { "B", "KB", "MB", "GB", "TB" };
        double v = b;
        int u = 0;
        while (v >= 1000 && u < units.Length - 1)
        {
            v /= 1000;
            ++u;
        }
        return u == 0 ? $"{b} B" : $"{v:0.0} {units[u]}";
    }

    public static string Eta(long seconds) => seconds switch
    {
        < 0 => "",
        < 60 => $"≈ {seconds} s",
        < 3600 => $"≈ {Math.Round(seconds / 60.0)} min",
        _ => $"≈ {seconds / 3600} h {seconds % 3600 / 60} min",
    };

    public static string Rate(double bytesPerSecond) =>
        bytesPerSecond <= 0 ? "" : $"{bytesPerSecond / 1_000_000:0} MB/s";
}
